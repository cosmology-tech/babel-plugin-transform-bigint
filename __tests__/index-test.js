// to run this test file use `npx jest` in the parent folder or `npm run test`
// to transform some-file.js use a command: `npx babel --plugins=module:./index.js some-file.js` from the parent folder

// src/__tests__/index-test.js
const babel = require('@babel/core');
const plugin = require('../index.js');


it('it works with AssignmentExpressions', function () {
  const example = `
    const o = {};
    o.x = {};
    o.x.y = 1n;
    o.x.yz = 1n;
    const y = 'y';
    const z = 'z';
    const b = 1n;
    const array = [1n];
    let i = 1;

    o.x.y += b;
    o.x['y'] += b;
    o.x[y] += b;
    o.x[y + z] += b;
    array[i] += b;
    array[0] += b;
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

/*
it('it works with UpdateExpression', function () {
  const example = `
    const o = {};
    o.x = {};
    o.x.y = 1n;
    o.x.yz = 1n;
    const y = 'y';
    const z = 'z';
    const b = 1n;

    ++o.x.y;
    ++o.x['y'];
    ++o.x[y];
    ++o.x[y + z];

    o.x.y++;
    o.x['y']++;
    o.x[y]++;
    o.x[y + z]++;
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});
*/

it('works when type of variable is changed', function () {
  const example = `
    let g1 = 1;
    g1 = 1n;
    if (g1 === 1n) {
      console.log(g1);
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('non-strict comparisions are not changed', function () {
  const example = `
    const g = 1n;
    if (g == 1) {
      console.log(g);
    }
    if (g != 1) {
      console.log(g);
    }
    if (g < 1) {
      console.log(g);
    }
    if (g > 1) {
      console.log(g);
    }
    if (g <= 1) {
      console.log(g);
    }
    if (g >= 1) {
      console.log(g);
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('BigInt.asUintN(n, a) is replaced', function () {
  const example = `
    const g = 1n;
    BigInt.asUintN(10, g)
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('works', function () {
  const example = `
    function f() {
      const x = 1n;
      return x + x;
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('BigInt.asUintN(64, a), BigInt.asIntN(64, a)', function () {
  const example = `
    const a = 1n;
    console.log(BigInt.asUintN(64, a));
    console.log(BigInt.asIntN(64, a));
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('typeof type guard (see https://www.typescriptlang.org/docs/handbook/2/narrowing.html#typeof-type-guards)', function () {
  const example = `
    function f1(a) {
      if (typeof a !== 'bigint') {
        throw new RangeError();
      }
      return a * a;
    }
    function f2(a, b) {
      if (typeof a !== 'bigint' || typeof b !== 'bigint') {
        throw new RangeError();
      }
      return a * b;
    }
    function f3(a) {
      if (typeof a !== 'number') {
        throw new RangeError();
      }
      return a * a;
    }
    function f4(a, b) {
      if (typeof a !== 'number' || typeof b !== 'number') {
        throw new RangeError();
      }
      return a * b;
    }
    function f5(a) {
      if (typeof a !== 'bigint') {
        void 0;
      }
      return a * a;
    }
    export {f1, f2, f3, f4, f5};
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('it does not replace expression for a mutable variable', function () {
  const example = `
    function f() {
      for (let i = 0; i < 10; i += 1) {
        console.log(i * i);
      }
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('it does not replace expression for a mutable variable2', function () {
  const example = `
    function f() {
      let i = 1;
      i = -i;
      return i * i;
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});


it('default values', function () {
  const example = `
    function f(y = unknown(), z = y * y) {
      if (typeof y !== 'bigint') {
        throw new RangeError();
      }
      return y * y;
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});


it('arguments in non-strict mode', function () {
  const example = `
    function func(a) {
      if (typeof a !== 'number') {
        throw new RangeError();
      }
      arguments[0] = 99n;
      console.log(a * a);
    }
    func(10); // prints 9801
  `;
  try {
    const {code} = babel.transform(example, {plugins: [plugin]});
    expect(code).toMatchSnapshot();
  } catch (error) {
    console.assert(error instanceof RangeError);
  }
});

it('eval in non-strict mode', function () {
  const example = `
    function func(a) {
      if (typeof a !== 'number') {
        throw new RangeError();
      }
      eval('a = 99n')
      console.log(a * a);
    }
    func(10); // prints 9801
  `;
  try {
    const {code} = babel.transform(example, {plugins: [plugin]});
    expect(code).toMatchSnapshot();
  } catch (error) {
    console.assert(error instanceof RangeError);
  }
});


it('sometimes type of conditional expression can be determined as JSBI', function () {
  const example = `
    function f(a) {
      const b = a % 3n === 0n ? 1n : 3n;
      return b * b;
    }
    export default f;
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('internal number function', function () {
  const example = `
    function f(a) {
      return a * a;
    }
    console.log(f(3));
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('internal bigint function', function () {
  const example = `
    function f(a) {
      return a * a;
    }
    console.log(f(3n));
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('CallExpression\'s type', function () {
  const example = `
    function n() {
      return 3;
    }
    console.log(n() * n());
    function b() {
      return 3n;
    }
    console.log(b() * b());
    function nb() {
      if (Math.random() < 0.5) {
        return 3;
      }
      return 3n;
    }
    console.log(nb() * nb());
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('maybeJSBI', function () {
  const example = `
    function f(a) {
      return Number(BigInt.asUintN(64, a));
    }
    console.log(f(3n));
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});
it('maybeJSBI2', function () {
  const example = `
    function f(a) {
      return Number(BigInt(a) < BigInt(0));
    }
    console.log(f(3n));
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});
it('maybeJSBI2a', function () {
  const example = `
    function f(a) {
      const x = BigInt(a) < BigInt(0);
      return Number(x);
    }
    console.log(f(3n));
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});
it('maybeJSBI3', function () {
  const example = `
    function f(a) {
      return Math.floor(a + a);
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});


it('maybeJSBI avoided for a FunctionExpression', function () {
  const example = `
    const f = function (x) {
      if (typeof x !== 'number') {
        throw new RangeError();
      }
      return x * x;
    }
  `;
  const {code} = babel.transform(example, {plugins: [plugin]});
  expect(code).toMatchSnapshot();
});

it('options jsbiLib jsbi', function () {
  const example = `
    let g1 = 1;
    g1 = 1n;
    if (g1 === 1n) {
      console.log(g1);
    }
  `;
  const {code} = babel.transform(example, {plugins: [[ plugin , {jsbiLib: "jsbi"}]]});
  expect(code).toMatchSnapshot();
});

it('options jsbiLib ./jsbi.js', function () {
  const example = `
    let g1 = 1;
    g1 = 1n;
    if (g1 === 1n) {
      console.log(g1);
    }
  `;
  const {code} = babel.transform(example, {plugins: [[ plugin , {jsbiLib: "./jsbi.js"}]]});
  expect(code).toMatchSnapshot();
});