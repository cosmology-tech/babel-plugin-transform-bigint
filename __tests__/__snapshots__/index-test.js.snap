// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigInt.asUintN(64, a), BigInt.asIntN(64, a) 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
const a = JSBI.BigInt(1);
console.log(JSBI.asUintN(64, a));
console.log(JSBI.asIntN(64, a));"
`;

exports[`BigInt.asUintN(n, a) is replaced 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
const g = JSBI.BigInt(1);
JSBI.asUintN(10, g);"
`;

exports[`CallExpression's type 1`] = `
"var maybeJSBI = {
  toNumber: function toNumber(a) {
    return typeof a === \\"object\\" ? JSBI.toNumber(a) : Number(a);
  },
  add: function add(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.add(a, b) : a + b;
  },
  subtract: function subtract(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.subtract(a, b) : a - b;
  },
  multiply: function multiply(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.multiply(a, b) : a * b;
  },
  divide: function divide(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.divide(a, b) : a / b;
  },
  remainder: function remainder(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.remainder(a, b) : a % b;
  },
  exponentiate: function exponentiate(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.exponentiate(a, b) : typeof a === \\"bigint\\" && typeof b === \\"bigint\\" ? new Function(\\"a\\", \\"b\\", \\"return a**b\\")(a, b) : Math.pow(a, b);
  },
  leftShift: function leftShift(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.leftShift(a, b) : a << b;
  },
  signedRightShift: function signedRightShift(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.signedRightShift(a, b) : a >> b;
  },
  bitwiseAnd: function bitwiseAnd(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseAnd(a, b) : a & b;
  },
  bitwiseOr: function bitwiseOr(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseOr(a, b) : a | b;
  },
  bitwiseXor: function bitwiseXor(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseXor(a, b) : a ^ b;
  },
  lessThan: function lessThan(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.lessThan(a, b) : a < b;
  },
  greaterThan: function greaterThan(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.greaterThan(a, b) : a > b;
  },
  lessThanOrEqual: function lessThanOrEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.lessThanOrEqual(a, b) : a <= b;
  },
  greaterThanOrEqual: function greaterThanOrEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.greaterThanOrEqual(a, b) : a >= b;
  },
  equal: function equal(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.equal(a, b) : a === b;
  },
  notEqual: function notEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.notEqual(a, b) : a !== b;
  },
  unaryMinus: function unaryMinus(a) {
    return typeof a === \\"object\\" ? JSBI.unaryMinus(a) : -a;
  },
  bitwiseNot: function bitwiseNot(a) {
    return typeof a === \\"object\\" ? JSBI.bitwiseNot(a) : ~a;
  }
};
import JSBI from \\"./jsbi.mjs\\";
function n() {
  return 3;
}
console.log(n() * n());
function b() {
  return JSBI.BigInt(3);
}
console.log(JSBI.multiply(b(), b()));
function nb() {
  if (Math.random() < 0.5) {
    return 3;
  }
  return JSBI.BigInt(3);
}
console.log(maybeJSBI.multiply(nb(), nb()));"
`;

exports[`default values 1`] = `
"var maybeJSBI = {
  toNumber: function toNumber(a) {
    return typeof a === \\"object\\" ? JSBI.toNumber(a) : Number(a);
  },
  add: function add(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.add(a, b) : a + b;
  },
  subtract: function subtract(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.subtract(a, b) : a - b;
  },
  multiply: function multiply(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.multiply(a, b) : a * b;
  },
  divide: function divide(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.divide(a, b) : a / b;
  },
  remainder: function remainder(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.remainder(a, b) : a % b;
  },
  exponentiate: function exponentiate(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.exponentiate(a, b) : typeof a === \\"bigint\\" && typeof b === \\"bigint\\" ? new Function(\\"a\\", \\"b\\", \\"return a**b\\")(a, b) : Math.pow(a, b);
  },
  leftShift: function leftShift(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.leftShift(a, b) : a << b;
  },
  signedRightShift: function signedRightShift(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.signedRightShift(a, b) : a >> b;
  },
  bitwiseAnd: function bitwiseAnd(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseAnd(a, b) : a & b;
  },
  bitwiseOr: function bitwiseOr(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseOr(a, b) : a | b;
  },
  bitwiseXor: function bitwiseXor(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseXor(a, b) : a ^ b;
  },
  lessThan: function lessThan(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.lessThan(a, b) : a < b;
  },
  greaterThan: function greaterThan(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.greaterThan(a, b) : a > b;
  },
  lessThanOrEqual: function lessThanOrEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.lessThanOrEqual(a, b) : a <= b;
  },
  greaterThanOrEqual: function greaterThanOrEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.greaterThanOrEqual(a, b) : a >= b;
  },
  equal: function equal(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.equal(a, b) : a === b;
  },
  notEqual: function notEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.notEqual(a, b) : a !== b;
  },
  unaryMinus: function unaryMinus(a) {
    return typeof a === \\"object\\" ? JSBI.unaryMinus(a) : -a;
  },
  bitwiseNot: function bitwiseNot(a) {
    return typeof a === \\"object\\" ? JSBI.bitwiseNot(a) : ~a;
  }
};
import JSBI from \\"./jsbi.mjs\\";
function f(y = unknown(), z = maybeJSBI.multiply(y, y)) {
  if (!(y instanceof JSBI)) {
    throw new RangeError();
  }
  return maybeJSBI.multiply(y, y);
}"
`;

exports[`internal bigint function 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  return JSBI.multiply(a, a);
}
console.log(f(JSBI.BigInt(3)));"
`;

exports[`internal number function 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  return a * a;
}
console.log(f(3));"
`;

exports[`it does not replace expression for a mutable variable 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f() {
  for (let i = 0; i < 10; i += 1) {
    console.log(i * i);
  }
}"
`;

exports[`it does not replace expression for a mutable variable2 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f() {
  let i = 1;
  i = -i;
  return i * i;
}"
`;

exports[`it works with AssignmentExpressions 1`] = `
"var _x, _x2, _x3, _x4, _y;
import JSBI from \\"./jsbi.mjs\\";
const o = {};
o.x = {};
o.x.y = JSBI.BigInt(1);
o.x.yz = JSBI.BigInt(1);
const y = 'y';
const z = 'z';
const b = JSBI.BigInt(1);
const array = [JSBI.BigInt(1)];
let i = 1;
_x = o.x, _x.y = JSBI.add(_x.y, b);
_x2 = o.x, _x2['y'] = JSBI.add(_x2['y'], b);
_x3 = o.x, _x3[y] = JSBI.add(_x3[y], b);
_x4 = o.x, _y = y + z, _x4[_y] = JSBI.add(_x4[_y], b);
array[i] = JSBI.add(array[i], b);
array[0] = JSBI.add(array[0], b);"
`;

exports[`maybeJSBI 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  return JSBI.toNumber(JSBI.asUintN(64, a));
}
console.log(f(JSBI.BigInt(3)));"
`;

exports[`maybeJSBI avoided for a FunctionExpression 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
const f = function (x) {
  if (typeof x !== 'number') {
    throw new RangeError();
  }
  return x * x;
};"
`;

exports[`maybeJSBI2 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  return Number(JSBI.lessThan(JSBI.BigInt(a), JSBI.BigInt(0)));
}
console.log(f(JSBI.BigInt(3)));"
`;

exports[`maybeJSBI2a 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  const x = JSBI.lessThan(JSBI.BigInt(a), JSBI.BigInt(0));
  return Number(x);
}
console.log(f(JSBI.BigInt(3)));"
`;

exports[`maybeJSBI3 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  return Math.floor(a + a);
}"
`;

exports[`non-strict comparisions are not changed 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
const g = JSBI.BigInt(1);
if (g == 1) {
  console.log(g);
}
if (g != 1) {
  console.log(g);
}
if (g < 1) {
  console.log(g);
}
if (g > 1) {
  console.log(g);
}
if (g <= 1) {
  console.log(g);
}
if (g >= 1) {
  console.log(g);
}"
`;

exports[`options jsbiLib ./jsbi.js 1`] = `
"import JSBI from \\"./jsbi.js\\";
let g1 = 1;
g1 = JSBI.BigInt(1);
if (JSBI.equal(g1, JSBI.BigInt(1))) {
  console.log(g1);
}"
`;

exports[`options jsbiLib jsbi 1`] = `
"import JSBI from \\"jsbi\\";
let g1 = 1;
g1 = JSBI.BigInt(1);
if (JSBI.equal(g1, JSBI.BigInt(1))) {
  console.log(g1);
}"
`;

exports[`sometimes type of conditional expression can be determined as JSBI 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f(a) {
  const b = JSBI.equal(JSBI.remainder(a, JSBI.BigInt(3)), JSBI.BigInt(0)) ? JSBI.BigInt(1) : JSBI.BigInt(3);
  return JSBI.multiply(b, b);
}
export default f;"
`;

exports[`typeof type guard (see https://www.typescriptlang.org/docs/handbook/2/narrowing.html#typeof-type-guards) 1`] = `
"var maybeJSBI = {
  toNumber: function toNumber(a) {
    return typeof a === \\"object\\" ? JSBI.toNumber(a) : Number(a);
  },
  add: function add(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.add(a, b) : a + b;
  },
  subtract: function subtract(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.subtract(a, b) : a - b;
  },
  multiply: function multiply(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.multiply(a, b) : a * b;
  },
  divide: function divide(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.divide(a, b) : a / b;
  },
  remainder: function remainder(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.remainder(a, b) : a % b;
  },
  exponentiate: function exponentiate(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.exponentiate(a, b) : typeof a === \\"bigint\\" && typeof b === \\"bigint\\" ? new Function(\\"a\\", \\"b\\", \\"return a**b\\")(a, b) : Math.pow(a, b);
  },
  leftShift: function leftShift(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.leftShift(a, b) : a << b;
  },
  signedRightShift: function signedRightShift(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.signedRightShift(a, b) : a >> b;
  },
  bitwiseAnd: function bitwiseAnd(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseAnd(a, b) : a & b;
  },
  bitwiseOr: function bitwiseOr(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseOr(a, b) : a | b;
  },
  bitwiseXor: function bitwiseXor(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.bitwiseXor(a, b) : a ^ b;
  },
  lessThan: function lessThan(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.lessThan(a, b) : a < b;
  },
  greaterThan: function greaterThan(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.greaterThan(a, b) : a > b;
  },
  lessThanOrEqual: function lessThanOrEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.lessThanOrEqual(a, b) : a <= b;
  },
  greaterThanOrEqual: function greaterThanOrEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.greaterThanOrEqual(a, b) : a >= b;
  },
  equal: function equal(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.equal(a, b) : a === b;
  },
  notEqual: function notEqual(a, b) {
    return typeof a === \\"object\\" && typeof b === \\"object\\" ? JSBI.notEqual(a, b) : a !== b;
  },
  unaryMinus: function unaryMinus(a) {
    return typeof a === \\"object\\" ? JSBI.unaryMinus(a) : -a;
  },
  bitwiseNot: function bitwiseNot(a) {
    return typeof a === \\"object\\" ? JSBI.bitwiseNot(a) : ~a;
  }
};
import JSBI from \\"./jsbi.mjs\\";
function f1(a) {
  if (!(a instanceof JSBI)) {
    throw new RangeError();
  }
  return JSBI.multiply(a, a);
}
function f2(a, b) {
  if (!(a instanceof JSBI) || !(b instanceof JSBI)) {
    throw new RangeError();
  }
  return JSBI.multiply(a, b);
}
function f3(a) {
  if (typeof a !== 'number') {
    throw new RangeError();
  }
  return a * a;
}
function f4(a, b) {
  if (typeof a !== 'number' || typeof b !== 'number') {
    throw new RangeError();
  }
  return a * b;
}
function f5(a) {
  if (!(a instanceof JSBI)) {
    void 0;
  }
  return maybeJSBI.multiply(a, a);
}
export { f1, f2, f3, f4, f5 };"
`;

exports[`works 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
function f() {
  const x = JSBI.BigInt(1);
  return JSBI.add(x, x);
}"
`;

exports[`works when type of variable is changed 1`] = `
"import JSBI from \\"./jsbi.mjs\\";
let g1 = 1;
g1 = JSBI.BigInt(1);
if (JSBI.equal(g1, JSBI.BigInt(1))) {
  console.log(g1);
}"
`;
